test/test_fdb_retrieve.f90

```diff
@@ -0,0 +1,116 @@
+program test_fdb_retrieve
+   use fdb
+   use eccodes
+
+   integer(c_int)                            :: res, err
+   type(c_ptr)                               :: fdb_handle
+   type(c_ptr)                               :: req ! Request 
+   type(c_ptr)                               :: dr ! DataReader
+   type(c_ptr)                               :: it ! ListIterator
+
+   integer(c_long)                           :: size, read, messageLength, marker
+
+   integer                                   :: i, ifile, igrib, iret
+   character(len=10)                         :: open_mode = 'r'
+
+   character(kind=c_char, len=1), dimension(:), allocatable  :: buf
+   character(kind=c_char, len=1), dimension(:), allocatable  :: message
+
+   CHARACTER(len=32)                         :: date_values(1)
+   CHARACTER(len=32)                         :: parameterNumber_values(2)
+   CHARACTER(len=32)                         :: single_value(1)
+
+   integer                                   :: msgid, status
+   character(len=128)                        :: keyvalue_str
+
+   integer                                   :: numberOfValues
+   real, dimension(:), target, allocatable   :: values
+
+   character(kind=c_char, len=1), dimension(100)  :: uri
+   integer(kind=c_int)                            :: off, len, max_len
+   logical(kind=c_bool)                           :: duplicates = .false.
+
+   integer :: idx_a, idx_b, next_message_idx, next_grib
+   ! grib api error messages
+   character(len=24) :: gribErrorMsg = 'Error reading grib file'
+   character(len=20) :: gribFunction = 'test'
+
+   res = fdb_initialise()
+   res = fdb_new_handle(fdb_handle)
+   res = fdb_new_request(req)
+
+   call fdb_request_add_fortran(req, "validityDate", ["20220703"])
+   call fdb_request_add_fortran(req, "validityTime", ["0"])
+   call fdb_request_add_fortran(req, "productionStatusOfProcessedData", ["255"])
+   call fdb_request_add_fortran(req, "productDefinitionTemplateNumber", ["0"])
+   parameterNumber_values(1)='0'
+   parameterNumber_values(2)='254'
+   call fdb_request_add_fortran(req, "parameterNumber", parameterNumber_values)
+   call fdb_request_add_fortran(req, "parameterCategory", ['0','1','3'])
+   call fdb_request_add_fortran(req, "discipline", ['0','2'])
+   call fdb_request_add_fortran(req, "generatingProcessIdentifier", ["153"])
+   call fdb_request_add_fortran(req, "typeOfFirstFixedSurface", ["sfc"])
+   call fdb_request_add_fortran(req, "level", ["0"])
+
+   res = fdb_new_datareader(dr);
+
+   res = fdb_list(fdb_handle, req, it, duplicates)
+
+   err = fdb_listiterator_next(it)
```

Comment:
you could add some comments in this process, since this example explains how to use the API. 
i.e. step to create the request, iterate over all message to extract the maximum len, etc.

Reply:
Added some comments.